lib/rucio/db/sqla/session.py

```diff
@@ -323,6 +323,7 @@ def new_funct(*args, **kwargs):
         except Exception:
             raise
     new_funct.__doc__ = function.__doc__
+    new_funct.__signature__ = signature(function)
```

Comment:
Is it possible to do this without `signature`? One idea I have is to use generic types like this, which maybe provides better IDE support:

```python3
from typing import TypeVar, Callable, cast

_F = TypeVar('_F', bound=Callable)

def read_session(function: _F) -> _F:
    ...
    def new_funct(...):
        ...
    return cast(_F, new_funct)
```

Edit: Saw the test for argspec further down, I'm guessing `dogpile` explicitly wants it this way then?

Reply:
Yes, dogpile is very explicit about that... xD
https://github.com/sqlalchemy/dogpile.cache/blob/1de93aab14c1274f20c1f44f8adff3b143c864f6/dogpile/util/compat.py#L33-L34